File: December 2024 Leetcode Solution/December-14.c
long long continuousSubarrays(int* nums, int numsSize) {
    int start = 0;
    long long totalSubarrays = 0;
    int* maxDeque = (int*)malloc(numsSize * sizeof(int));
    int* minDeque = (int*)malloc(numsSize * sizeof(int));
    int maxFront = 0, maxBack = 0;
    int minFront = 0, minBack = 0;

    for (int end = 0; end < numsSize; end++) {
        while (maxBack > maxFront && nums[maxDeque[maxBack - 1]] <= nums[end]) {
            maxBack--;
        }
        maxDeque[maxBack++] = end;
        while (minBack > minFront && nums[minDeque[minBack - 1]] >= nums[end]) {
            minBack--;
        }
        minDeque[minBack++] = end;
        while (nums[maxDeque[maxFront]] - nums[minDeque[minFront]] > 2) {
            if (maxDeque[maxFront] == start) {
                maxFront++;
            }
            if (minDeque[minFront] == start) {
                minFront++;
            }
            start++;
        }
        totalSubarrays += (end - start + 1);
    }

    free(maxDeque);
    free(minDeque);
    return totalSubarrays;
}

File: December 2024 Leetcode Solution/December-14.java
class Solution {
    public long continuousSubarrays(int[] nums) {
        int start = 0;
        long totalSubarrays = 0;
        Deque<Integer> maxDeque = new LinkedList<>();
        Deque<Integer> minDeque = new LinkedList<>();

        for (int end = 0; end < nums.length; end++) {
            while (!maxDeque.isEmpty() && nums[maxDeque.peekLast()] <= nums[end]) {
                maxDeque.pollLast();
            }
            maxDeque.offerLast(end);
            while (!minDeque.isEmpty() && nums[minDeque.peekLast()] >= nums[end]) {
                minDeque.pollLast();
            }
            minDeque.offerLast(end);
            while (nums[maxDeque.peekFirst()] - nums[minDeque.peekFirst()] > 2) {
                if (maxDeque.peekFirst() == start) {
                    maxDeque.pollFirst();
                }
                if (minDeque.peekFirst() == start) {
                    minDeque.pollFirst();
                }
                start++;
            }
            totalSubarrays += (end - start + 1);
        }

        return totalSubarrays;
    }
}
